File: 1_to_100/7/solution.py


def merge_two_sorted_arrays(num_list1,n1,num_list2,n2):
    # num_list1 is the first array contianing sorted 
    # array and extra zeros for the elements of second array
    # n1 is the number of elements first array
    # n2 is the number of elements second array

    # special case for one of two arrays having no elements
    if n1 == 0:
        return num_list2
    if n2 == 0:
        return num_list1

    # for this problem we can start filling the first array from end 
    # the size of num_list1 is n1+n2
    while n1>0 and n2>0:
        # get the max elements of two 
        if num_list1[n1-1]> num_list2[n2-1]:
            num_list1[n1+n2-1]=num_list1[n1-1]
            n1 = n1 -1 
        else: 
            num_list1[n1+n2-1] = num_list2[n2-1]
            n2 = n2 -1
    if n2 != 0:
        num_list1[n1+n2-1] = num_list2[n2-1]
    return num_list1

if __name__ == "__main__":
    print(merge_two_sorted_arrays([1,2,3,0,0,0],3,[2,5,6],3))

File: 1_to_100/7/solution.cpp
#include <iostream>
#include <bits/stdc++.h>

using namespace std;

void merge_two_sorted_arrays(vector<int>& nums1, int m,vector<int>& nums2,int n){
    // nums1 is sorted and of size m+n, first m elements are original elements 
    // of the array other elements are zeros (places for second array elements) 
    // nums2 is sorted and of size n
    while(m>=0||n>=0){
        if(m>0 & n>0){
            // set the max element of the two arrays
            if(nums1[m-1] > nums2[n-1]){
                nums1[n+m-1] = nums1[m-1];
                m--;
            }else{

                nums1[n+m-1] = nums2[n-1];
                n--;
            }
        }else{
            cout<<n<<" "<<m<<endl;
            if (n==0){
                break;
            }else{
                nums1[n+m-1] = nums2[n-1];
                n--;
            }
        }
    }
}

vector<int> array_to_vector(int array[], int n){
    vector<int> result;
    for(int i=0; i<n; i++){
        result.push_back(array[i]);
    }
    return result;
}

int main(){
    int x[] = {0};
    int x_size = 1;
    vector<int> nums1 = array_to_vector(x,x_size);
    int y[] = {5};
    int y_size = 1;
    vector<int> nums2 = array_to_vector(y,y_size);
    merge_two_sorted_arrays(nums1,x_size-y_size,nums2, y_size);
    for(int i=0; i<nums1.size(); i++){
        cout << nums1[i]<< endl;
    }
}